homeassistant/components/digital_ocean/__init__.py

```diff
@@ -81,6 +89,31 @@ def get_droplet_id(self, droplet_name):
 
         return droplet_id
 
+    @Throttle(MIN_TIME_BETWEEN_DOMAIN_UPDATES)
+    def update_domain_record(
+        self, domain_name, record_name, record_value, record_type="A"
+    ):
+        """Update the appointed DNS record with the desired value."""
+        domain = digitalocean.Domain(token=self._access_token, name=domain_name)
+        try:
+            records = domain.get_records()
+        except digitalocean.baseapi.NotFoundError as exc:
+            raise DomainRecordsNotFound(
```

Comment:
Is there a reason why this is wrapped in a custom exception?

Reply:
No particular reason, other than conveying semantical context about what is happening.
Not that this exception is used in two different locations:
1. When the domain you requested to change does not exist on your DigitalOcean account
2. When the domain exists, but the requested record does not.

Both cases result in the same outcome, which is the record was not updated
So I want to convey to the caller (`handle_update_domain_record`) what matters for it:
- Was the record updated?
- If not, why?

Reply:
> No particular reason

That was the point for asking... as in, we can just handle those errors in the placed used at this point? Right now it is captures, wrapped in a new one, raise and captured again?... feels like unneeded complexity added.

Reply:
Not really. Right now it's one custom exception that encapsulates two different errors:
- one that is wrapping another exception.
- one that is purely semantical

I think this is merely a trade-off: where do we want exception-handling logic? Do we want it on this function, or do we want it on the handler?

IMO this is better, as it:
1. Encompasses two different errors with the same outcome
2. Hides library-specific errors from the service handler
3. Conveys a good semantic to the reader: 
    Calling that function either succeeds or produces ONE exception whose name accurately describes the issue

Reply:
> I think this is merely a trade-off: where do we want exception-handling logic? Do we want it on this function, or do we want it on the handler?

That is not a trade-off, it is now in 2 places... 🤷 

With that in mind I do not agree on the points you make after that. "Conveys a good semantic to the reader" is very personal. To me, it adds unneeded complexity I need to learn and follow, for no reason whatsoever.

This begs the question of why we need this method at all (and why isn't this handled/done by the service handler directly).

Reply:
> To me, it adds unneeded complexity I need to learn and follow, for no reason whatsoever.

I disagree. I'd rather learn to follow and keep basic encapsulation principles:
See line [72-73](https://github.com/home-assistant/core/pull/89800/files#diff-1a40e66a06bceb5ed78b1ded6f19ae5eb77587205319c8589ca4769418fbd26cR72-R73): The `DigicalOcean` class is responsible for all access to the external service (not my decision). I merely kept the pattern.

Now, on readability, my new method in that class has two very clear outcomes. It either:
- Successfully executes the update
- Raises an exception (which now inherits from `HomeAssistantError`)

As a utility method from an "API wrapper", this seems pretty OK:
It's up to the caller (a service handler, a sensor, or whatever) to decide what to do with that exception.

Consider the alternatives:
1. Moving all the logic to the service handler would leak API access outside of that class. The service handler would need to extract the API key from the settings and do the dirty work
2. Moving all of the code to the existing class would leak service logic to a utility class that otherwise needs to know nothing about "services" or "sensors"


At last, being in two places is indeed not ideal, but IMO it's better to have that service handler in another file than keep it on the (already crowded) `__init__.py` file... If anything, I would move the `DigitalOcean` class to yet another file.

Let me know your thoughts on this.